### DISCUSSION TITLE
Edit in Adobe Suite - middleware idea

### DISCUSSION THREAD

**Discussion by cfc62** [RhetTbull/osxphotos#1404] [Ideas]
I haven't been an Adobe fan for some time, but at this point in time they're offering photo editing tools that other's either don't - or can't - and they seem to be increasing their lead. I've bought RAW Power and Photomator but they just aren't cutting it at this point. Yes, the RAW editing is there but for any sort of repair using generative fill, there's nothing that compares with Adobe. 

At the same time, the Photos ecosystem is going to remain a great way of consuming content. I do not want to leave that ecosystem.

I do really like how RAW Power and Photomator are integrated right into Photos. You choose to edit in one of these tools, and the changes you make are in your Photos library with no muss, no fuss. 

I do not see any such integration with Adobe products such as Lightroom or Photoshop (hey, if I've missed it, please let me know).

(And yes, I'm aware that there's significant product overlap here.)

So the idea is this:

Middleware that supports:

1. Selecting a single photo in Photos
2. Telling Photos I want to edit in another app (this functionality is already clearly supported and in use by RAW Power and Photomator)
3. Handing over that photo to the Adobe suite/open that asset in Adobe
4. .....edit in Adobe app(s)...
5. Hand the asset back to Photos when done

This would give the best of both worlds - keep all of the assets in Photos for collection and consumption, but allow for sophisticated editing in the Adobe suite. 

What I've found for this process is complicated and burdensome. Yes, it can be done but it's not a smooth process and files get duplicated and converted. No, I do not want to convert a RAW file into a monstrous 16-bit TIFF so Adobe can read it, when I know full well Adobe can already read all RAW file formats. Then you have to get it back into Photos, and that's another process.

I haven't looked at Adobe for a while (years) so I don't know what APIs they have out there these days; perhaps this isn't even possible but I wanted to get this idea out there for discussion.

Hoping that constraining this to one file at a time is helpful; of course Adobe is never going to support the Photos library schema but if that can be worked around in middleware that would be the key. No GUI, no UI, just a choice in the "Edit with..." menu selection for one photo.

**Comment by RhetTbull** [comment]
I don't use Adobe products (I prefer the Affinity Photo suite which is more affordable and offers everything I need) so can't comment on any integrations. Photos does offer the ability to [edit with an external editor](https://support.apple.com/en-us/102259#:~:text=Control%2Dclick%20a%20photo%20and,party%20app%20that%20you%20chose.) -- does this not work with  Photoshop?

I do have a [crazy idea](https://github.com/RhetTbull/osxphotos/discussions/1233) for allowing an osxphotos export to have bi-directional sync with Photos. Basically, a photo edited in the export would be detected as an edit to the original and synced back to Photos as an edit.  I think it's technically feasible but it's not something I have time to work on at the moment. One regret would be that you'd lose any metadata about the edit so lose ability to do lossless edits/undo, etc. Currently, when editing a photo in Photos or in a external editor that supports integration with Photos, the editor can save a sidecar with metadata about the edit (it's either an .aae or .plist file) that can then be used to undo the edits. You'd lose that with the approach I'm thinking about but you'd gain the flexibility of being able to use the library in Photos or out of photos. With APFS formatted disk, the standard on all modern Macs, osxphotos already uses "copy on write" which means the exported copy doesn't take extra disk space (until edited) so this comes basically for "free" in terms of storage.

**Comment by cfc62** [comment]
I'm struggling with this, TBH. Note, I'm only talking about Lightroom Classic, I'm not using the Cloud version. 

Yes, you can edit photos in _Photoshop_ via the "Edit with..." selection. This seems to work to some level so I can get generative fill & repair, which is a good chunk of what I'm looking for. I need to experiment with this more. 

But not _Lightroom_. Well, not with RAW. It does open LR Classic but only the JPEG makes it over, and I don't yet understand how (or even if) to save my work so it gets back into Photos. Without the RAW portion I'm unsure how to make this work reasonably.

I suppose I could simply point Lightroom at my originals part of Photos and have it create a completely parallel library, but that just seems to be ripe for disastrous error. 

What's more likely is that I go through my library and make a new album of photos that I want to edit in Adobe, then export those, then open them up as a new library in Lightroom Classic, then do whatever I need to do (Lightroom hands over to Photoshop fairly well), then (sigh) re-import the edited files back into Photos.

**Reply by RhetTbull** [reply]
It might be possible to create a small middleware app that's a native [Photos extension](https://developer.apple.com/documentation/photokit/creating_photo_editing_extensions?language=objc) that would pop up  a dialog to ask you about sending the RAW or JPEG to Lightroom then handle the round trip. I'll need to think of this. If so it wouldn't be part of osxphotos but a standalone app.

**Reply by cfc62** [reply]
A few observations and points of confusion:

- Somewhere between Photos and Adobe, the "Edit with Lightroom" got nuked. Doesn't really matter as it was only sending over jpgs anyways
- I've found I'm not the only person struggling with this overall. 
- There are library conversion programs (Avalanche) but there are serious drawbacks (e.g. many edits aren't directly supported and are faked using AI, with predictably poor results)
- Yes, Adobe will ingest your Photos library but all edits are lost. Great. And you can only do it once.
- HDR is a mess. Getting HDR photos out of the Apple ecosystem is a challenge (Photomator has added functionality, but it doesn't look 100%)
- Adobe doesn't really have any consumption apps that compare to Apple's
- Adobe's suite is **_crazy_** RAM- and CPU-hungry. Even editing one RAW can send it to outer space. Plus, it adds a ton of cloud-related services which run even if you disallow them. Crappy programming. 

In the end though, Adobe has photo editing tools that aren't matched. RAW Power and Photomator and such do work within Photos, but they lack the functionality that Adobe brings. In addition, you can find a lot of help with Adobe apps, but little to none for the others - including Photos. 

Bottom line, I'd be thrilled with something that let me work on a selected Photos asset in Adobe, with the results coming  back to my Photos library. I have not found such a thing yet.

**Reply by RhetTbull** [reply]
Tell me a little more about workflow. For example, Photos only supports storing the "edited" version in JPEG but it does provide an option for storing a sidecar with data about the edits so the edited version can be constructed from the original by the program that did the edits. What does editing a raw file in Adobe do? Does it modify the raw file? Does it save data in a sidecar? What about editing a JPEG? Is it stored in .PS format after editing? If there was a way to figure out exactly what Adobe needed to reconstruct the edits, a JPEG could be stored as expected by Photos and the other data could be kept in the sidecar to pass back to Adobe when needed.

**Reply by cfc62** [reply]
I'll do some experimenting to see what happens, but I do know one (unusable) method that comes into play is a 16-bit TIFF; these get up into the several hundreds of MBs. 

In general, most of these apps do non-destructive edits and the edit files are saved in various formats and in different ways, and they vary from app to app (which is why Avalanche exists). 

But I'll try to hunt down specifics. Might be a few days.

**Reply by cfc62** [reply]
That sound you heard was me opening a huge can of worms... only to find another can of worms contained within. 

Part - but only part - of the issue here in learning about files/formats/capabilities is that the landscape is changing rapidly, and what was accurate and definitive even just a year ago is incorrect today, especially in the realm of HDR. 

Some of the HDR support in Photos and the rest of the Apple ecosystem - including Safari -  has only been added (quietly) in the last few months, but it's still not complete. We know how silent Apple can be about these sort of things, so when the complete chain might be HDR enabled is anyone's guess. Things were very confusing as I read through articles which confidently said HDR wasn't yet supported in this that or the other app, only to find out that it had been subsequently added and the article invalidated. 

There are also new file formats coming that will natively support HDR, but as I understand it, current file formats have been made to support HDR as well, even JPG via tone mapping. 

Why the emphasis on HDR when I only mentioned it as one of several items? Well, I found out that all of my RAW images (and I've been shooting RAW+JPEG for a while now) make for great HDR. I didn't know that at the time, and now I realize I'm sitting on thousands of images that can go from OK to pretty good or even stunning just by unlocking the HDR potential. 

My curiosity was also piqued when I made a smart album in Photos of only HDR pictures, and it - somehow, still haven't figured this out - includes regular JPGs taken by iPhones as old as the iPhone 5. Is this an error? I don't recall Apple supporting HDR until iPhone 10 or so. 

Also found out that ProRAW uses Adobe's very well established and open-standard RAW format DNG - it's not another proprietary RAW format, nicely enough. Apple did add some extensions to the DNG format but to their credit Adobe incorporated them (at least that's my understanding). 

I'm nowhere near done understanding what's going on here overall, but I'm very pleased that I've got a trove of RAW files convertible to HDR. I've already looked at a few in Lightroom Classic and once you select the HDR view the vibrance comes through. I don't know if or how to get that HDR goodness back into the Apple ecosystem so I can view these on my HDR TV with an Apple TV, which is how I "consume" my images - there's no Adobe equivalent for consumption. 

At the same time, I still want to be able to get AI generative fill and repair for images with distracting elements, as well as more refined editing tools. Adobe does that quite well today, if I'm reading the tea leaves correctly I believe Apple will have some of this capability with the next release, but that's a guess. Adobe has it now. 

I very reluctantly bought another external SSD yesterday to support exporting my Photos catalog to an off board Lightroom library, or at least to experiment with it. It's still not what I want to do, but the siren call of unlocking HDR and fixing images in the Adobe suite is just too strong at the moment. 

Not to be underestimated is the vast amount of support and knowledge in the Adobe community that unfortunately just doesn't exist on the Photos side. Almost all of the accurate and insightful knowledge and tips that I've picked up in the last few days has come from the Adobe camp. Interestingly, I got more information on Photos from the Adobe community than anywhere else. 

Sigh. But it's good to have helpful resources, regardless of the source. 

I'm still working on this.
